CMakeLists.txt

```diff
@@ -0,0 +1,178 @@
+    async_grpc/rpc.cc
+    async_grpc/event_queue_thread.cc)
+
+set(ALL_TESTS
+    async_grpc/server_test.cc
+    async_grpc/type_traits_test.cc)
+
+list(REMOVE_ITEM ALL_LIBRARY_SRCS ${ALL_TESTS})
+
+set(INSTALL_SOURCE_HDRS ${ALL_LIBRARY_HDRS})
+
+file(GLOB_RECURSE ALL_PROTOS "async_grpc*/*.proto")
+set(ALL_PROTOS
+    async_grpc/proto/math_service.proto)
+
+set(ALL_PROTO_SRCS)
+set(ALL_PROTO_HDRS)
+foreach(REL_FIL ${ALL_PROTOS})
+  get_filename_component(DIR ${REL_FIL} DIRECTORY)
+  get_filename_component(FIL_WE ${REL_FIL} NAME_WE)
+
+  list(APPEND ALL_PROTO_SRCS "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.cc")
+  list(APPEND ALL_PROTO_HDRS "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.h")
+
+  add_custom_command(
+    OUTPUT "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.cc"
+           "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.h"
+    COMMAND  ${PROTOBUF_PROTOC_EXECUTABLE}
+    ARGS --cpp_out  ${PROJECT_BINARY_DIR} -I
+      ${PROJECT_SOURCE_DIR} ${REL_FIL}
+    DEPENDS ${REL_FIL}
+    COMMENT "Running C++ protocol buffer compiler on ${REL_FIL}"
+    VERBATIM
+  )
+endforeach()
+set_source_files_properties(${ALL_PROTO_SRCS} ${ALL_PROTO_HDRS} PROPERTIES GENERATED TRUE)
+list(APPEND ALL_LIBRARY_HDRS ${ALL_PROTO_HDRS})
+list(APPEND ALL_LIBRARY_SRCS ${ALL_PROTO_SRCS})
+
+add_library(${PROJECT_NAME} ${ALL_LIBRARY_HDRS} ${ALL_LIBRARY_SRCS})
+
+foreach(REL_FIL ${ALL_TESTS})
+  get_filename_component(DIR ${REL_FIL} DIRECTORY)
+  get_filename_component(FIL_WE ${REL_FIL} NAME_WE)
+  # Replace slashes as required for CMP0037.
+  string(REPLACE "/" "." TEST_TARGET_NAME "${DIR}/${FIL_WE}")
+  google_test("${TEST_TARGET_NAME}" ${REL_FIL})
+  target_link_libraries("${TEST_TARGET_NAME}" PUBLIC grpc++)
+endforeach()
+
+target_link_libraries(${PROJECT_NAME} PUBLIC glog)
+target_link_libraries(${PROJECT_NAME} PUBLIC gflags)
+
+target_include_directories(${PROJECT_NAME} SYSTEM PUBLIC
+  ${PROTOBUF_INCLUDE_DIR})
+# TODO(hrapp): This should not explicitly list pthread and use
+# PROTOBUF_LIBRARIES, but that failed on first try.
+target_link_libraries(${PROJECT_NAME} PUBLIC ${PROTOBUF_LIBRARY} pthread)
+target_link_libraries(${PROJECT_NAME} PUBLIC grpc++)
+
+# Add the binary directory first, so that port.h is included after it has
+# been generated.
+target_include_directories(${PROJECT_NAME} PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}>
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
+    $<INSTALL_INTERFACE:include>
+)
+
+# TODO(damonkohler): Create a testing library.
+target_include_directories(${PROJECT_NAME} SYSTEM PRIVATE
+  "${GMOCK_INCLUDE_DIRS}")
+target_link_libraries(${PROJECT_NAME} PRIVATE ${GMOCK_LIBRARY})
+
+set(TARGET_COMPILE_FLAGS "${TARGET_COMPILE_FLAGS} ${GOOG_CXX_FLAGS}")
+set_target_properties(${PROJECT_NAME} PROPERTIES
+  COMPILE_FLAGS ${TARGET_COMPILE_FLAGS})
+
+install(
+  TARGETS ${PROJECT_NAME}
+  EXPORT AsyncGrpcExport
+  ARCHIVE DESTINATION lib
+  LIBRARY DESTINATION lib
+  RUNTIME DESTINATION bin
+)
+
+foreach(HDR ${INSTALL_SOURCE_HDRS})
```

Comment:
list append INSTALL_GENERATED_HDRS, then you can drop the next block

Reply:
generated headers is empty anyways. I am dropping it.